pennylane/ops/qubit/parametric_ops.py

```diff
@@ -3028,3 +3028,144 @@ def adjoint(self):
+
+    Args:
+        phi (float): the phase angle
+        wires (int): the subsystem the gate acts on
+        do_queue (bool): Indicates whether the operator should be
+            immediately pushed into the Operator queue (optional)
+        id (str or None): String representing the operation (optional)
+    """
+    num_wires = 2
+    num_params = 1
+    """int: Number of trainable parameters that the operator depends on."""
+
+    grad_method = "A"
+    grad_recipe = ([[0.5, 1, np.pi / 2], [-0.5, 1, -np.pi / 2]],)
+
+    def __init__(self, phi, wires, do_queue=True, id=None):
+        super().__init__(phi, wires=wires, do_queue=do_queue, id=id)
+
+    @staticmethod
+    def compute_decomposition(phi, wires):
+        r"""Representation of the operator as a product of other operators (static method). :
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+
+        .. seealso:: :meth:`~.PSWAP.decomposition`.
+
+        Args:
+            phi (float): the phase angle
+            wires (Iterable, Wires): the subsystem the gate acts on
+
+        Returns:
+            list[Operator]: decomposition into lower level operations
+
+        **Example:**
+
+        >>> qml.PSWAP.compute_decomposition(1.23, wires=(0,1))
+        [SWAP(wires=[0, 1]), CNOT(wires=[0, 1]), PhaseShift(1.23, wires=[1]), CNOT(wires=[0, 1])]
+        """
+        return [
+            qml.SWAP(wires=wires),
+            qml.CNOT(wires=wires),
+            qml.PhaseShift(phi, wires=[wires[1]]),
+            qml.CNOT(wires=wires),
+        ]
+
+    @staticmethod
+    def compute_matrix(phi):  # pylint: disable=arguments-differ
+        r"""Representation of the operator as a canonical matrix in the computational basis (static method).
+
+        The canonical matrix is the textbook matrix representation that does not consider wires.
+        Implicitly, this assumes that the wires of the operator correspond to the global wire order.
+
+        .. seealso:: :meth:`~.PSWAP.matrix`
+
+
+        Args:
+           phi (tensor_like or float): phase angle
+
+        Returns:
+           tensor_like: canonical matrix
+
+        **Example**
+
+        >>> qml.PSWAP.compute_matrix(0.5)
+        array([[1.        +0.j, 0.        +0.j        , 0.        +0.j        , 0.        +0.j],
+              [0.        +0.j, 0.        +0.j        , 0.87758256+0.47942554j, 0.        +0.j],
+              [0.        +0.j, 0.87758256+0.47942554j, 0.        +0.j        , 0.        +0.j],
+              [0.        +0.j, 0.        +0.j        , 0.        +0.j        , 1.        +0.j]])
+        """
+        if qml.math.get_interface(phi) == "tensorflow":
+            phi = qml.math.cast_like(phi, 1j)
+
+        e = qml.math.exp(1j * phi)
+
+        return qml.math.stack(
+            [
+                stack_last([1, 0, 0, 0]),
+                stack_last([0, 0, e, 0]),
+                stack_last([0, e, 0, 0]),
+                stack_last([0, 0, 0, 1]),
+            ],
+            axis=-2,
+        )
```

Comment:
I don't remember anymore, but I think it was mentioned somewhere to use `stack`.

Reply:
Could you point out where it was mentioned? @antalszava maybe you know more about this.

Reply:
It's been so long; I don't remember it anymore.

Reply:
Hmm, wasn't sure either, after digging with the help of `git blame` it seems there were some cases [originally where `qml.math.array` was not working as desired with differentiability](https://github.com/PennyLaneAI/pennylane/pull/1749#discussion_r729157018) :thinking: Stacking seems to have been introduced in https://github.com/PennyLaneAI/pennylane/pull/1749, (pinging @glassnotes and @josh146 if they have some extra info here).

Reply:
I don't remember the exact reason (I believe it was actually Josh who discovered that it was required!). But I think it was a mix of typecasting, differentiability, and also ensuring consistency across different interfaces in terms of the shape of the input parameter.

Reply:
Yep! Exactly as @glassnotes mentioned. Creating an array directly with specific elements does not work as expected on all ML frameworks. On some, it will cast to a NumPy array. On others, it might still create a tensor, but you will lose differentiability.

Instead, you need to recast array construction as a set of differentiable functions. E.g., you could

- use `scatter` to take particular tensors, and the indices that they are to placed at, and create an array
- use `stack` to stack each row, and then stack all the rows together
- etc. (there are probably many other approaches, each with trade offs).

While `stack` was used here because it was conceptually easy to code (and also works well with broadcasting since you can pass an `axis` argument) there might be a faster way of doing this.

Reply:
Good to know, thanks!